components/calendar/src/julian.rs

```diff
@@ -0,0 +1,198 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+//! This module contains types and implementations for the Julian calendar
+
+#[derive(Copy, Clone, Debug, Default)]
+// The Julian calendar
+pub struct Julian;
+
+#[derive(Copy, Clone, Debug, Hash, Eq, PartialEq)]
+// The inner date type used for representing Date<Georgian>
+pub struct JulianDateInner(IsoDateInner);
+
+impl Calendar for Julian {
+    type DateInner = JulianDateInner;
+    fn date_from_iso(&self, iso: Date<Iso>) -> JulianDateInner {
+        *iso.inner()
+    }
+
+    fn date_to_iso(&self, date: &Self::DateInner) -> Date<Iso> {
+        Date::from_raw(*date, Iso)
+    }
+
+    fn months_in_year(&self, date: &Self::DateInner) -> u8 {
+        Iso.months_in_year(&date.0)
+    }
+
+    fn days_in_year(&self, date: &Self::DateInner) -> u32 {
+        if Self::is_leap_year(date.year) {
+            366
+        } else {
+            365
+        }
+    }
+
+    fn days_in_month(&self, date: &Self::DateInner) -> u8 {
+        Self::days_in_month(date.year, date.month);
+    }
+
+    fn day_of_week(&self, date: &Self::DateInner) -> types::IsoWeekday {
+        let leap_years_offset = date.year.0 / 4;
+        let days_to_current_year = 365 * date.year.0 + leap_years;
+        let year_offset = days_to_current_year % 7;
+
+        let month_offset = if Self::is_leap_year(date.year) {
```

Comment:
Given that the _only_ thing different between these functions is `is_leap_year`, i wonder if some of this code can be shared on ISODateInner and have it be parametrized over `F: Fn(Year) -> bool` for the `is_leap_year` bit.

Reply:
This would make it trivial to reimplement offset as well

Reply:
It's actually probably better to generalize the offset function, and for weekday just convert to ISO first. This could probably be a default implementation on the `Calendar` trait, even.

Reply:
Does making leap_year as a trait that has to be implemented for Calendar make sense?

Reply:
No, because other calendars do this differently. I think for now we should make the _functions_ generic, and think about unifying further later.

For example, the Indian National (Shaka) calendar probably can share some code as well: It's a calendar _like_ the Gregorian/Julian ones but has different month lengths.

However the Islamic and Hebrew calendars work totally differently and cannot share much code. Perhaps the offset function can be shared but it will not be as easy.

Reply:
My general hope here is that we incrementally make things more generic: So here we should make the offset/until functions sufficiently generic to support ISO and Julian (but not other calendars, and Gregorian/Buddhist can rely on ISO), but not further. Then when we add the Indian National Calendar we should figure out how to make `offset()` and `until()` more generic.

Reply:
Ok, for now, just copying it from iso and reimplementing it for julian

Reply:
Sounds good.